### PR TITLE
Add JSON parser

### PR DESCRIPTION
**Pull request by dekrain**
Still need some work.

### CODE REVIEW

**Comment by Anonymus-Player** [comment]
oh, i didn't know it was here =))

File: cpp/StringView.hpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+        {
+            if (start >= _size)
+                return out_of_range{};
+
+            const usize max_length = _size - start;
+            return basic_string_view(_data + start, math::min(len, max_length));
+        }
+
+        // I/O
+        friend std::basic_ostream<CharT>& operator<<(std::basic_ostream<CharT>& os, basic_string_view v) {
```
Comment (Anonymus-Player) [inline cpp/StringView.hpp:119]: Are you using `iostream` because it's quicker?

File: cpp/Result.hpp [hunk 2]
```diff
@@ -424,6 +429,11 @@ namespace hsd
             return _initialized;
         }
 
+        explicit constexpr operator bool() const
```
Comment (Anonymus-Player) [inline cpp/Result.hpp:432]: there is a `.is_ok()`, reason?

File: cpp/Json.hpp [hunk 3]
```diff
@@ -0,0 +1,405 @@
+#pragma once
+
+/// @brief JSON stream parser
+
+#include <queue>
```
Comment (Anonymus-Player) [inline cpp/Json.hpp:5]: If I recall correctly queues couldn't be implemented using linked lists

File: cpp/Json.hpp [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+                            current_token = JsonToken::True;
+                            token_kw = s_keywords[1];
+                            ++token_position;
+                            break;
+                        case static_cast<CharT>('f'):
+                            current_token = JsonToken::False;
+                            token_kw = s_keywords[2];
+                            ++token_position;
+                            break;
+                        case static_cast<CharT>('"'):
```
Comment (Anonymus-Player) [inline cpp/Json.hpp:101]: You didn't need to do `'\"'` if I'm right?

File: cpp/Json.hpp [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+        }
+
+        // If an error occurs, the rest of buffer can be passed in
+        Result<void, JsonError> lex(vstr frag) {
+            static const char* const s_keywords[] = {"null", "true", "false"};
+            for (CharT ch : frag) {
+                ++pos;
+                _reparse:
+                if (current_token == JsonToken::Empty) {
+                    if (std::isspace(ch))
```
Comment (Anonymus-Player) [inline cpp/Json.hpp:70]: Explain this?

File: cpp/Json.hpp [hunk 6]
```diff
@@ -0,0 +1,405 @@
+#pragma once
+
+/// @brief JSON stream parser
+
+#include <queue>
+#include <cstring>
```
Comment (Anonymus-Player) [inline cpp/Json.hpp:6]: Checking if a character is a space is the only thing you use from there

File: cpp/StringView.hpp (hunk 1 above)
Comment (dekrain) [reply cpp/StringView.hpp:119]: Yes, I'll probably switch to our I/O when/or before I finish the parser

File: cpp/Result.hpp (hunk 2 above)
Comment (dekrain) [reply cpp/Result.hpp:432]: To use in `if`s and boolean expressions. It's just quicker.

File: cpp/Json.hpp (hunk 4 above)
Comment (dekrain) [reply cpp/Json.hpp:101]: No, quotes need to be escaped only if surrounded by matching quotes.

File: cpp/Json.hpp (hunk 5 above)
Comment (dekrain) [reply cpp/Json.hpp:70]: To skip whitespace, duh

File: cpp/Json.hpp (hunk 6 above)
Comment (dekrain) [reply cpp/Json.hpp:6]: I'll try to integrate this into `CString` but I may need to lookup how Unicode does that

File: cpp/Json.hpp (hunk 5 above)
Comment (Anonymus-Player) [reply cpp/Json.hpp:70]: You couldn't do `if (ch == static_cast<CharT>(' '))`?

Comment (dekrain) [reply cpp/Json.hpp:70]: Do you know how many whitespace characters are there?

Comment (dekrain) [reply cpp/Json.hpp:70]: m a n y

File: Tests/Json/JsonTest.cpp [hunk 7]
```diff
@@ -8,7 +8,7 @@ LR"json({
     "Cow": "milk",
     "gold": 5000,
     "primes": [2, 3, 5, 7, 11, 13],
-    "Unicode": "蛋",
+    "Unicode": "‏蛋‎‎‎‎‎‎‎‎‎‎‎‎‎‎‎",
```
Comment (Anonymus-Player) [inline Tests/Json/JsonTest.cpp:11]: how is this a different character?

**Review by Anonymus-Player — COMMENTED** [review]
Can be used to parse files?

File: Tests/Json/JsonTest.cpp (hunk 7 above)
Comment (dekrain) [reply Tests/Json/JsonTest.cpp:11]: ¯\_(ツ)_/¯ ¯\_(ツ)_/¯ ¯\_(ツ)_/¯ :slightly_smiling_face:

**Comment by dekrain** [comment]
> Can be used to parse files?

Yes of course, just call `lexer.lex(contents).unwrap()` then `lex.push_eot().unwrap()`

**Comment by Anonymus-Player** [comment]
> > Can be used to parse files?
> 
> Yes of course, just repeatedly call `lexer.lex(chunk).unwrap()` for each `chunk` of a file, and then `lexer.push_eot().unwrap()`
Maybe you should write that in the tests and make a more detailed guide about this particular feature

**Pull request merged**